File: SinaNewsSearchAnalysis/src/main/java/offline/crawler/NewsCrawler.java
package offline.crawler;

import offline.utils.Category;
import offline.utils.DatabaseUtil;
import org.apache.http.HttpEntity;
import org.apache.http.NameValuePair;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.client.utils.URIBuilder;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClients;
import org.apache.http.message.BasicNameValuePair;
import org.apache.http.util.EntityUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.json.JSONArray;
import org.json.JSONObject;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.io.IOException;
import java.net.URISyntaxException;
import java.sql.SQLException;
import java.util.*;


public class NewsCrawler {

    // 日志记录
    static final Logger logger = LogManager.getLogger(NewsCrawler.class.getName());

    // 请求地址
    final String newsUrl = "http://roll.news.sina.com.cn/interface/rollnews_ch_out_interface.php";

    // 爬虫连接
    CloseableHttpClient httpClient;

    // 存储一些公共的请求头信息
    List<NameValuePair> nvps;

    // 数据库操作相关
    DatabaseUtil db;

    public NewsCrawler() {
        logger.info("爬虫初始化...");
        httpClient = HttpClients.createDefault();
        nvps = new ArrayList<NameValuePair>();
        // 请求参数
        //col=89&spec=&type=&ch=01&k=&offset_page=0&offset_num=0&num=60&asc=&page=1
        nvps.add(new BasicNameValuePair("col", "89"));
        nvps.add(new BasicNameValuePair("spec", ""));
        nvps.add(new BasicNameValuePair("type", ""));
        nvps.add(new BasicNameValuePair("ch", "01"));
        nvps.add(new BasicNameValuePair("k", ""));
        nvps.add(new BasicNameValuePair("offset_page", "0"));
        nvps.add(new BasicNameValuePair("offset_num", "0"));
        nvps.add(new BasicNameValuePair("num", "60"));
        nvps.add(new BasicNameValuePair("asc", ""));
        // page 需要动态调整，因为要爬取多页的内容

        // 初始化一个Database 对象
        db = new DatabaseUtil();
        logger.info("数据连接完毕！");

    }

    public void run(int start, int end) throws URISyntaxException, IOException, SQLException {
        for (int i = start; i <= end; ++i) {
            HttpGet httpGet = new HttpGet(newsUrl);
            URIBuilder uriBuilder = new URIBuilder(httpGet.getURI()).addParameters(nvps);
            uriBuilder.addParameter("page", String.valueOf(i));
            httpGet.setURI(uriBuilder.build());
            // System.out.println(uriBuilder);
            httpGet.setHeader("User-Agent",
                    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/63.0.3239.132 Safari/537.36");
            httpGet.setHeader("Referer", "http://roll.news.sina.com.cn/s/channel.php?ch=01");
            httpGet.setHeader("Accept-Language", "zh-CN,zh;q=0.9,en;q=0.8,zh-TW;q=0.7");
            httpGet.setHeader("Accept-Encoding", "gzip, deflate");
            httpGet.setHeader("Host", "roll.news.sina.com.cn");

            CloseableHttpResponse response = httpClient.execute(httpGet);
            response.setHeader("Content-Type", "text/html");

            HttpEntity entity = response.getEntity();
            // System.out.println(response.getStatusLine());
            if (entity != null) {
                // 日志记录
                logger.info("爬取第 " + i + " 页的滚动新闻...");
                // 中文乱码解决方案：http://blog.csdn.net/pkuyjxu/article/details/8712220
                String gbk = new String(EntityUtils.toString(entity).getBytes("8859_1"), "gb2312");
                // System.out.println("response content: " + gbk2utf8);
                // System.out.println(gbk);
                gbk = gbk.replaceAll("\\\\\"", "");
                gbk = gbk.replaceAll("\\\\", "");
                gbk = gbk.substring(gbk.indexOf('{'));
                // System.out.println(gbk);
                String gbk2utf8 = new String(gbk.getBytes("utf-8"), "utf-8");
                JSONObject obj = new JSONObject(gbk2utf8);
                JSONArray array = obj.getJSONArray("list");
                //System.out.println(array);
                for(int j = 0; j<array.length(); ++j) {
                    NewsModel nm = new NewsModel();
                    JSONObject jobj = array.getJSONObject(j);
                    // System.out.println(jobj);
                    String category =jobj.getJSONObject("channel").getString("title");
                    nm.setCategory(Category.cate2Int(category));
                    String title = jobj.getString("title");
                    nm.setTitle(title);
                    String url = jobj.getString("url");
                    if (!db.checkUrl(url)) {
                        nm.setLink(url);
                        Integer time = jobj.getInt("time");
                        nm.setTime(time);
                        // System.out.println(nm);
                        TitleResult tr = getContent(url);
                        if (tr.getContent().length() > 1) {
                            nm.setContent(tr.getContent());
                            if (tr.getTitle().length() > 1) {
                                title = tr.getTitle();
                                nm.setTitle(title);
                            }
                            logger.info("成功获取 《" + title + "》新闻内容！");
                            db.insertNews(nm);
                        }
                        try {
                            Thread.sleep(new Random().nextInt(10000));
                        } catch (InterruptedException e) {
                            e.printStackTrace();
                        }
                    }
                    else {
                        logger.warn("新闻已存在于数据库中，跳过");
                    }
                }
            }
            // 释放资源，避免卡死
            httpGet.releaseConnection();
            response.close();
        }
        // 程序结束后关闭连接
        closeHttpClient();

    }

    private void closeHttpClient() throws IOException {
        httpClient.close();
    }

    public TitleResult getContent(String url) throws IOException {
        //CloseableHttpClient httpclient = HttpClients.createDefault();
        HttpGet httpGet = new HttpGet(url);
        httpGet.setHeader("User-Agent",
                "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/63.0.3239.132 Safari/537.36");
        httpGet.setHeader("Referer", "http://roll.news.sina.com.cn/s/channel.php?ch=01");
        httpGet.setHeader("Accept-Language", "zh-CN,zh;q=0.9,en;q=0.8,zh-TW;q=0.7");
        httpGet.setHeader("Accept-Encoding", "gzip, deflate");
        httpGet.setHeader("Connection", "keep-alive");

        CloseableHttpResponse response = httpClient.execute(httpGet);

        // System.out.println(response.getStatusLine());
        TitleResult tr = new TitleResult();
        String result = "";
        String html = "html";
        HttpEntity entity = response.getEntity();

        if (entity != null) {
            // logger.info(url);
            html = EntityUtils.toString(entity, "utf-8");
            Document doc = Jsoup.parse(html);
            if (doc.select("h1.main-title").first() != null) {
                tr.setTitle(doc.select("h1.main-title").first().text());
            }
            else {
                tr.setTitle("");
            }
            Element article = doc.select("div.article").first();
            if (article == null) {
                article = doc.selectFirst("div#artibody");
            }
            if (article != null) {
                Elements paras = article.select("p");
                if (paras != null) {
                    ArrayList<String> paraList = new ArrayList<String>();
                    for (Element e : paras) {
                        String line = e.text().trim().replace("　", "");
                        if (line.length() > 1)
                            paraList.add(line);
                    }
                    result = String.join("\n", paraList.toArray(new String[paraList.size()]));
                    tr.setContent(result);
                }
            }
        }
        // 释放资源，避免卡死
        response.close();
        httpGet.releaseConnection();
        return tr;
    }

    class TitleResult {
        private String title;
        private String content;

        public TitleResult() {
            title = new String();
            content = new String();
        }

        public TitleResult(String title, String content) {
            this.title = title;
            this.content = content;
        }

        public String getTitle() {
            return title;
        }

        public void setTitle(String title) {
            this.title = title;
        }

        public String getContent() {
            return content;
        }

        public void setContent(String content) {
            this.content = content;
        }
    }
}

File: NewsAbExtraction/newsCrawler/newsCrawler.py
import json
import pprint
import random
import time
import urllib
import requests

from bs4 import BeautifulSoup
from snownlp import SnowNLP

# 一些全局变量
# 模拟浏览器的访问的信息
from databaseUtil import DatabaseUtil

headers = {
    'Content-Type': 'text/html',
    'User-Agent': "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/63.0.3239.132 Safari/537.36",
    'Referer': "http://roll.news.sina.com.cn/s/channel.php?ch=01",
    'Accept-Language': "zh-CN,zh;q=0.9,en;q=0.8,zh-TW;q=0.7",
    'Accept-Encoding': "gzip, deflate",
    'Host': "roll.news.sina.com.cn",
}
# 获取新闻的入口
newsUrl = "http://roll.news.sina.com.cn/interface/rollnews_ch_out_interface.php"
# 获取新闻链接中需要的一些参数
globalParams = {
  'col': '89',
  'spec': '',
  'type': '',
  'ch': '01',
  'k': '',
  'offset_page': '0',
  'offset_num': '0',
  'num': '60',
  'asc': '',
}

# 帮助解析 JSON 的辅助类
class Dummy(dict):
  def __getitem__(self, item):
    return item

class NewsCrawler():
  def __init__(self):
    # 初始化一个数据库对象
    self.dbUtil = DatabaseUtil()

  # 根据新闻的 url，获取新闻的具体内容
  def getNewsContent(self, url: str):
    content = []
    req = urllib.request.Request(url=url)
    response = urllib.request.urlopen(req, timeout=5)
    try:
      response_read = response.read().decode('utf-8')
    except:
      response_read = response.read().decode('gbk')
    soup = BeautifulSoup(response_read, 'lxml')
    article = soup.select_one('div.article')
    if not article:
      article = soup.select_one('div#artibody')
    paras = article.select('p')
    if paras:
      for para in paras:
        line = para.text.strip()
        if len(line) >= 1:
          content.append(line)
    return content

  # Python3 支持的指定参数形式
  def parseNewsDetail(self, r: dict):
    entity = {}
    # 新闻标题
    entity['title'] = r['title']
    print(entity['title'])
    # 新闻的 url，进一步获取新闻的具体内容，从而生成关键字、摘要
    entity['url'] = r['url']
    print(entity['url'])
    # 新闻的时间
    timeInt = int(r['time'])
    entity['time'] = timeInt
    # 新闻的具体文本内容
    content = self.getNewsContent(entity['url'])
    contentConcat = ''
    if len(content) >= 1:
      contentConcat = '\n'.join(content)
    entity['content'] = contentConcat
    # 使用 SnowNLP 提供的生成关键词、提取摘要的接口
    # 具体算法是 TextRank，毕业论文需要具体介绍这种方法
    if contentConcat:
      s = SnowNLP(contentConcat)
      # 生成关键词，因为有的时候关键词是一个单字（常用字），所以只保留长度大于等于 2 的
      keywords10 = s.keywords(10)
      keywords = [x for x in keywords10 if (len(x) > 1 and self.validKeywords(x))]
      keywordsSQL = '|'.join(keywords)
      entity['keywords'] = keywordsSQL
      # 生成一个三句话的摘要
      summary = s.summary(3)
      summarySQL = '|'.join(summary)
      entity['summary'] = summarySQL
      # print(keywordsSQL)
      # print(summarySQL)
      # print('='*80)
      # 写入数据库中
      self.dbUtil.insert(entity)

  def getSinaRollNews(self, startPage, endPage):
    for i in range(startPage, endPage):
      print("Page {}".format(i))
      params = globalParams.copy()
      params['page'] = str(i)
      allHtml = requests.get(newsUrl, params=params, headers=headers)
      pageHtml = allHtml.content.decode('gbk')
      pageHtml = pageHtml[pageHtml.index('{'):-1]
      # 解析获取到的 JSON 格式的新闻列表
      data_str = eval(pageHtml, Dummy())
      data_str = json.dumps(data_str)
      data_str = json.loads(data_str)
      data_str = data_str['list']
      for r in data_str:
        # 对每一个具体的新闻都进行解析，并且写入数据库
        self.parseNewsDetail(r)
        # 稍微暂停一下，避免爬取频率太高
        time.sleep(random.random()*6.0)
      time.sleep(10)

  # 有时候得到的关键词是一些标点符号、纯数字等，所以过滤掉
  def validKeywords(self, x: str):
    if ('—' in x) or ('：') in x or ('"' in x) or \
        ('，' in x) or ('。' in x) or (x.isdigit()):
      return False
    else:
      return True

if __name__ == '__main__':
  nc = NewsCrawler()
  # 获取页面[a,b)，不包括第 b 页，自己设置
  nc.getSinaRollNews(1, 5)
  # 下面是测试一下能否从数据库中查找到新闻
  # db = DatabaseUtil()
  # db.selectNewsByKeywords('今天')
  # res = db.selectNewsByKeywords('也许，真确，美国')
  # for x in res:
    # pprint.pprint(x)
  # rs = db.selectNewsById(1)
  # pprint.pprint(rs)


